apps/trading/components/deposit-container/use-squid-route.ts

```diff
@@ -0,0 +1,192 @@
+}) => {
+  const [queryKey, setQueryKey] = useState<FormFields>();
+  const { data: squid } = useSquid();
+
+  // List to changes to all fields (debounced) and store into some state
+  // that we can use as the query key for the route
+  useEffect(() => {
+    const callback = debounce((fields) => {
+      // Check against the schema so that the queryKey is not set if the form
+      // is invalid, therefore avoiding a failed fetch attempt for a swap route
+      if (formSchema.safeParse(fields).success) {
+        setQueryKey(fields);
+      } else {
+        setQueryKey(undefined);
+      }
+    }, 700);
+
+    const subscription = form.watch((x) => callback(x));
+
+    return () => subscription.unsubscribe();
+  }, [form]);
+
+  const queryResult = useQuery({
+    queryKey: ['squidRoute', queryKey],
+    enabled: enabled && Boolean(queryKey),
+    queryFn: async ({ queryKey }) => {
+      const queryKeyRes = formSchema.safeParse(queryKey[1]);
+      if (!queryKeyRes.success) return null;
+      if (!squid) return null;
+
+      const fields = queryKeyRes.data;
+
+      const fromAsset = squid.tokens.find(
+        (t) => t.address === fields.fromAsset && t.chainId === fields.fromChain
+      );
+
+      if (!fromAsset) return null;
+      if (!toAsset) return null;
+
+      // From token and to token are the same, not a valid swap
+      if (
+        fromAsset.address.toLowerCase() ===
+        toAsset.source.contractAddress.toLowerCase()
+      ) {
+        return null;
+      }
+
+      const config = configs.find((c) => c.chain_id === toAsset.source.chainId);
+
+      if (!config) return null;
+
+      const fromAmount = removeDecimal(fields.amount, fromAsset.decimals);
+
+      // The default bridgeAddress for the selected toAsset if an arbitrum
+      // to asset is selected will get changed to the squid receiver address
+      let bridgeAddress = config.collateral_bridge_contract
+        .address as `0x${string}`;
+      let approveCallData;
+      let depositCallData;
+
+      // Squid cannot guarantee that the funds will make it to the end desitination
+      // so there is a squid reciever contarct we should use if we are making a swap
+      // into an arbitrum asset
+      //
+      // TODO: there will in future be an ethereum squid receiver contract. We will
+      // need to ensure that contracts is used if making a swap to an Ethereum token
+      if (toAsset.source.chainId === ARBITRUM_CHAIN_ID) {
+        bridgeAddress = ARBITRUM_SQUID_RECEIVER_ADDRESS;
+        approveCallData = encodeFunctionData({
+          abi: getErc20Abi({
+            address: toAsset.source.contractAddress as `0x${string}`,
+          }),
+          functionName: 'approve',
+          args: [bridgeAddress, BigInt(fromAmount)],
+        });
+
+        // NOTE that the squid receiver deposit contract takes a 4th argument which is the recovery
+        // address if squid is unable to fulfill the swap
+        depositCallData = encodeFunctionData({
```

Comment:
suggestion: we have `encodeDepositData` and `encodeApproveData` on the contracts class that could be reused here

Reply:
I decided to switch to using wagmi/viem helpers rather that our wrappers that still use ethers.